File: test/src/com/leftrightfold/test/trixx.clj
;;  The contents of this file are subject to the Mozilla Public License
;;  Version 1.1 (the "License"); you may not use this file except in
;;  compliance with the License. You may obtain a copy of the License at
;;  http://www.mozilla.org/MPL/
;;
;;  Software distributed under the License is distributed on an "AS IS" 
;;  basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the 
;;  License for the specific language governing rights and limitations
;;  under the License.
;;
;;  The Original Code is Trixx.
;;
;;  The Initial Developer of the Original Code is Aaron Feng
;;  Portions created by Aaron Feng are Copyright (C) Aaron Feng 2009.
;;
;;  All Rights Reserved.
;;
;;  Contributor(s): ______________________________________.

(defn test-add-delete-and-list-queues []
  (assert (zero? (count (list-queues "/"))))
  (add-queue "/" "guest" "guest" "my-queue" true)
  (assert (= 1 (count (list-queues "/"))))
  (delete-queue  "/" "guest" "guest" "my-queue")
  (assert (zero? (count (list-queues "/")))))

(defn test-add-delete-and-list-exchanges []
  ; 7 default exchanges ships with rabbit by default
  (assert (= 7 (count (list-exchanges "/"))))
  (add-exchange "/" "guest" "guest" "my-exchange" "direct" true)
  (assert (= 8 (count (list-exchanges "/"))))
  (delete-exchange  "/" "guest" "guest" "my-exchange")
  (assert (= 7 (count (list-exchanges "/")))))

(defn test-add-delete-and-list-bindings []
  (assert (zero? (count (list-bindings "/"))))
  (assert (add-queue "/" "guest" "guest" "my-queue" true))
  (assert (add-exchange "/" "guest" "guest" "my-exchange" "direct" true))
  (assert (add-binding "/" "guest" "guest" "my-queue" "my-exchange" "my-key"))
  ; when creating a queue, a default binding is created without routing key.
  (assert (= 2 (count (list-bindings "/"))))
  (assert (delete-exchange  "/" "guest" "guest" "my-exchange"))
  (assert (delete-queue  "/" "guest" "guest" "my-queue")))

(defn test-add-delete-and-list-vhosts []
  ; one for default '/' vhost
  (assert (= 1 (count (list-vhosts))))
  (assert (add-vhost "my-vhost"))
  (assert (= 2 (count (list-vhosts))))
  (assert (delete-vhost "my-vhost")))

(defn test-add-delete-and-list-users []
  ; default 'guest'
  (assert (= 1 (count (list-users))))
  (assert (add-user "my-user" "password"))
  (assert (= 2 (count (list-users))))
  (assert (delete-user "my-user")))

(defn test-set-clear-permissions []
  ; when the user is first created, it is not associted with vhost until permissions are set
  ; ?how do you query for that user?
  (assert (add-user "my-user" "password"))
  (assert (set-permissions "my-user" "/" {:config "c" :write "w" :read "r"}))
  ;;; needs to find my-user
  
  (let [user (first (list-user-permissions "my-user"))]
    (assert (= "c" (:config user)))
    (assert (= "w" (:write user)))
    (assert (= "r" (:read user))))

  (assert (clear-permissions "my-user" "/"))
  (assert (list-user-permissions "my-user"))
  (assert (delete-user "my-user")))

(defn monitor-suite []
  (assert (stop-app))
  (assert (reset))
  (assert (start-app))
  (test-add-delete-and-list-queues)
  (test-add-delete-and-list-exchanges)
  (test-add-delete-and-list-bindings)
  (test-add-delete-and-list-vhosts)
  (test-add-delete-and-list-users)
  ;;; test list-connections
  (test-set-clear-permissions))

(monitor-suite)

File: src/com/leftrightfold/web/trixx-web/app/models/trixx.rb
require 'httparty'

class Trixx
  include HTTParty
  default_params :output => 'json'
  format :json
  base_uri "http://localhost:8080"
  
  def self.authenticate(name, password)
    result = post("/sessions/authenticate", :query => { :name     => name,
                                                        :password => password})
    result.code == 200
  end

  def self.stop
    result = put("/rabbit/stop")
    result.code == 200
  end

  def self.start
    result = put("/rabbit/start")
    result.code == 200
  end

  def self.reset
    result = put("/rabbit/reset")
    result.code == 200
  end
	
	def self.status
    Status.new(get("/rabbit/status"))
  end
  
  def self.exchanges
    get("/exchanges").collect do |exchange_hash|
      Exchange.new(exchange_hash)
    end
  end
	
  def self.queues
    get("/queues").collect do |queue_hash|
      Queue.new(queue_hash)
    end
  end	

  def self.bindings
    get("/bindings").collect do |binding_hash|
      Binding.new(binding_hash)
    end
  end
  
  def self.connections
    get("/connections").collect do |connection_hash|
      Connection.new(connection_hash)
    end
  end
  
  def self.users
    get("/users").collect do |user_hash|
      User.new(user_hash)
    end
  end 
  
  def self.update_user_atrributes(attributes)
    result = put("/users/#{attributes[:name]}", :query => { :name              => name, 
                                                            :vhost             => vhost, 
                                                            :config_permission => config_permission,
                                                            :write_permission  => write_permission,
                                                            :read_permission   => read_permission })
    User.new(attributes) if result.code == 200
  end

  def self.find_user_by_name(name) 
    # should throw an exception if result.code is not 200
    result = get("/users/#{URI.escape(name)}")
    User.new(result.to_hash) if result.code == 200 
  end

  def self.delete_user(name)
    # should throw an exception if result.code is not 200
    result = delete("/users/#{URI.escape(name)}")

    true if result.code == 200
  end

  def self.add_user(user)
    # should throw an exception if result.code is not 200
    result = post("/users", :query => { :name      => user.name, 
                                        :password  => user.password, 
                                        :vhost     => user.vhost, 
                                        :config    => user.config,
                                        :write     => user.write,
                                        :read      => user.read })
    result.code == 200
  end
end
